File: tasks/java-rate-limiter-001/starter/RateLimiter.java
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

public final class RateLimiter {
  private static final class Bucket {
    double tokens;
    Instant last;
    Bucket(double tokens, Instant last) { this.tokens = tokens; this.last = last; }
  }

  private final int capacity;
  private final double refillPerSecond;
  private final Map<String, Bucket> buckets = new HashMap<>();

  public RateLimiter(int capacity, double refillPerSecond) {
    this.capacity = capacity;
    this.refillPerSecond = refillPerSecond;
  }

  public boolean allow(String key, Instant now) {
    // TODO: refill the key's bucket, consume one token, and reject empty buckets.
    return false;
  }
}

File: tasks/csharp-rate-limiter-001/starter/RateLimiter.cs
using System;
using System.Collections.Generic;

public sealed class RateLimiter
{
    private sealed class Bucket
    {
        public double Tokens;
        public DateTimeOffset Last;

        public Bucket(double tokens, DateTimeOffset last)
        {
            Tokens = tokens;
            Last = last;
        }
    }

    private readonly int capacity;
    private readonly double refillPerSecond;
    private readonly Dictionary<string, Bucket> buckets = new();

    public RateLimiter(int capacity, double refillPerSecond)
    {
        this.capacity = capacity;
        this.refillPerSecond = refillPerSecond;
    }

    public bool Allow(string key, DateTimeOffset now)
    {
        // TODO: refill this key from elapsed time, consume one token, and
        // reject an empty bucket without sharing state with another key.
        return false;
    }
}
